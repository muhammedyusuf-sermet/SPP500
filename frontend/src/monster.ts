export enum Size {
	Tiny = "Tiny",
	Small = "Small",
	Medium = "Medium",
	Large = "Large",
	Huge = "Huge",
	Gargantuan = "Gargantuan"
}

export enum MonsterType {
	Beast = "Beast",
	Dragon = "Dragon",
	Monstrosity = "Monstrosity",
	Humanoid = "Humanoid",
	Fiend = "Fiend",
	Undead = "Undead",
	Giant = "Giant",
	Elemental = "Elemental",
	SwarmOfTinyBeasts = "SwarmOfTinyBeasts",
	Construct = "Construct",
	Celestial = "Celestial",
	Aberration = "Aberration",
	Fey = "Fey",
	Plant = "Plant",
	Ooze = "Ooze"
}

export enum MonsterRace {
	AnyRace = "AnyRace",
	Devil = "Devil",
	Demon = "Demon",
	Human = "Human",
	Shapechanger = "Shapechanger",
	Goblinoid = "Goblinoid",
	Titan = "Titan",
	Gnoll = "Gnoll",
	Gnome = "Gnome",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Orc = "Orc",
	Kobold = "Kobold",
	Lizardfolk = "Lizardfolk",
	Merfolk = "Merfolk",
	Sahuagin = "Sahuagin",
	Grimlock = "Grimlock"
}

export enum Alignment {
	Unaligned = "Unaligned",
	AnyAlignment = "AnyAlignment",
	AnyNonGoodAlignment = "AnyNonGoodAlignment",
	AnyNonEvilAlignment = "AnyNonEvilAlignment",
	AnyNonLawfulAlignment = "AnyNonLawfulAlignment",
	AnyNonChaoticAlignment = "AnyNonChaoticAlignment",
	AnyGoodAlignment = "AnyGoodAlignment",
	AnyEvilAlignment = "AnyEvilAlignment",
	AnyLawfulAlignment = "AnyLawfulAlignment",
	AnyChaoticAlignment = "AnyChaoticAlignment",
	AnyNeutralAlignment = "AnyNeutralAlignment",
	LawfulGood = "LawfulGood",
	LawfulNeutral = "LawfulNeutral",
	LawfulEvil = "LawfulEvil",
	NeutralGood = "NeutralGood",
	Neutral = "Neutral",
	NeutralEvil = "NeutralEvil",
	ChaoticGood = "ChaoticGood",
	ChaoticNeutral = "ChaoticNeutral",
	ChaoticEvil = "ChaoticEvil"
}

export enum Environment {
	Arctic = "Arctic",
	Coastal = "Coastal",
	Desert = "Desert",
	Forest = "Forest",
	Grassland = "Grassland",
	Hill = "Hill",
	Mountain = "Mountain",
	Swamp = "Swamp",
	Underdark = "Underdark",
	Underwater = "Underwater",
	Urban = "Urban"
}

export enum MonsterAction {
	SpecialAbility = "SpecialAbility",
	Action = "Action",
	LegendaryAction = "LegendaryAction"
}

export interface IMonsterAbilities {
	name: string,
	description: string,
}

export enum TypeOfAction {
	Ranged = "Ranged",
	Melee = "Melee",
}

export interface IMonsterActions {
	name: string,
	description: string,
	usageCount: number,
	legendaryAction: boolean,
	typeOfAction?: TypeOfAction,
	toHit: number,
	reach: number,
	targets: number,
	averagedDamage: number,
	damageDice: number,
	damageConstant: number,

}

export interface IMonsterState {
	Id?: number,
	Name: string,
	Size?: Size | string;
	Type?: MonsterType | string;
	Race?: MonsterRace | string;
	Environment?: Environment | string;
	Alignment?: Alignment | string;

	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;

	Speed?: string;
	Languages?: string;

	DamageVulnerabilities?: string;
	DamageResistances?: string;
	DamageImmunities?: string;
	ConditionImmunities?: string;

	ChallengeRating?: number;

	AbilityScores: IMonsterAbilityScoreState;
	Senses: SenseMap;
	Skills: SkillMap;
	SavingThrows: IMonsterSavingThrowState;
	//Actions: IActionState[];
}

export interface SkillMap {
	[skillName: string]: number|undefined;
}

export interface SenseMap {
	[senseName: string]: number|undefined;
}

export interface IMonsterAbilityScoreState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IMonsterSavingThrowState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IActionState {
	Name: string,
	Description: string,
	HitBonus?: number,
	Damage?: string,
	DamageBonus?: number,
	Type?: MonsterAction
}
/*export interface IMonster {
	name: string,
	type: MonsterType,
	alignment: MonsterAlignment,
	size: MonsterSize,
	race: MonsterRace,
	environment: MonsterEnvironment,
	resistance?: string,
	damageImmunity?: string,
	conditionImmunity?: string,
	vulnerability?: string,
	armorClass: number,
	hitPoints?: number,
	hitPointDice?: string,
	hitPointDiceAdd?: number,
	speedLand: number,
	speedSwim?: number,
	strStat: number,
	dexStat: number,
	conStat: number,
	intStat: number,
	wisStat: number,
	chaStat: number,
	strSavingThrow?: number,
	dexSavingThrow?: number,
	conSavingThrow?: number,
	intSavingThrow?: number,
	wisSavingThrow?: number,
	chaSavingThrow?: number,
	skillsAthletics: number,
	skillsAcrobatics?: number,
	skillsSleightOfHand?: number,
	skillsStealth?: number,
	skillsArcana?: number,
	skillsHistory?: number,
	skillsInvestigation?: number,
	skillsNature?: number,
	skillsReligion?: number,
	skillsAnimalHandling?: number,
	skillsInsight?: number,
	skillsMedicine?: number,
	skillsPerception?: number,
	skillsSurvival?: number,
	skillsDeception?: number,
	skillsIntimidation?: number,
	skillsPerformance?: number,
	skillsPersuasion?: number,
	sensesBlindsight?: number,
	sensesDarkvision?: number,
	sensesTremorsense?: number,
	sensesTruesight?: number,
	sensesPassivePerception?: number,
	sensesPassiveInvestigation?: number,
	sensesPassiveInsight?: number,
	languages?: string,
	challengeRating: number,
	experiencePoints: number,
	abilities?: IMonsterAbilities[],
	actions?: IMonsterActions[],
}*/