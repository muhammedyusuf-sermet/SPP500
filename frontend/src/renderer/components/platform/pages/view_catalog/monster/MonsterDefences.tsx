import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

export interface IMonsterDefencesProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		ArmorClass?: number;
		HitPoints?: number;
		HitPointDistribution?: string;
	}
}

export interface IMonsterDefencesState {
	// data
	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;
	// errors
	ArmorClassError?: string;
	HitPointsError?: string;
	HitPointDistributionError?: string;
}

export class MonsterDefences extends React.Component<IMonsterDefencesProps, IMonsterDefencesState> {
	constructor(props: IMonsterDefencesProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterDefencesProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false) {
			this.setState({
				...nextProps.initial
			});
		} else if (this.props.disabled != nextProps.disabled) {
			this.forceUpdate();
		}
	}

	stringToNumber = (toConvert : string) => {
		return isNaN(parseInt(toConvert)) ? undefined : parseInt(toConvert);
	}

	handleNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = this.stringToNumber(event.target.value);
		const name = event.currentTarget.name;
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	handleStringChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		let value: string|undefined = event.target.value;
		if (/^\s*$/.test(value))
			value = undefined
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [event.currentTarget.name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[event.currentTarget.name]: value,
					[event.currentTarget.name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	render() {
		return (
			<ExpansionPanel defaultExpanded CollapseProps={{timeout: 100}} >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Armor Class and Hit Points</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						<Grid item xs={12}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="ArmorClass">Armor Class</InputLabel>
								<Input
									id='ArmorClass'
									type='number'
									value={this.state.ArmorClass != undefined ? this.state.ArmorClass : ''}
									name='ArmorClass'
									onChange={this.handleNumberChange}
									aria-describedby="ArmorClass-helper-text" />
								<FormHelperText error id="ArmorClass-helper-text">{this.state.ArmorClassError}</FormHelperText>
							</FormControl>
						</Grid>
						<Grid item xs={6}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="HitPoints">Hit Points</InputLabel>
								<Input
									id='HitPoints'
									type='number'
									value={this.state.HitPoints != undefined ? this.state.HitPoints : ''}
									name='HitPoints'
									onChange={this.handleNumberChange}
									aria-describedby="HitPoints-helper-text" />
								<FormHelperText error id="HitPoints-helper-text">{this.state.HitPointsError}</FormHelperText>
							</FormControl>
						</Grid>
						<Grid item xs={6}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="HitPointDistribution">Hit Point Distribution</InputLabel>
								<Input
									id='HitPointDistribution'
									type='text'
									value={this.state.HitPointDistribution || ''}
									name='HitPointDistribution'
									onChange={this.handleStringChange}
									aria-describedby="HitPointDistribution-helper-text" />
								<FormHelperText error id="HitPointDistribution-helper-text">{this.state.HitPointDistributionError}</FormHelperText>
							</FormControl>
						</Grid>
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}