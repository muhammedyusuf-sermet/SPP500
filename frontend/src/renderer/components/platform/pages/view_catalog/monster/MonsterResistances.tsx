import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';
import '../../../../../css/app.css';

import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid, Tooltip } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import {Help} from "bloomer";

export interface IMonsterResistancesProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		DamageVulnerabilities?: string;
		DamageResistances?: string;
		DamageImmunities?: string;
		ConditionImmunities?: string;
	}
}

export interface IMonsterResistancesState {
	// data
	DamageVulnerabilities?: string;
	DamageResistances?: string;
	DamageImmunities?: string;
	ConditionImmunities?: string;
	// errors
	DamageVulnerabilitiesError?: string;
	DamageResistancesError?: string;
	DamageImmunitiesError?: string;
	ConditionImmunitiesError?: string;
}

export class MonsterResistances extends React.Component<IMonsterResistancesProps, IMonsterResistancesState> {
	constructor(props: IMonsterResistancesProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterResistancesProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false)
			this.setState({
				...nextProps.initial
			});
	}

	handleResistancesChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		let value: string|undefined = event.target.value;
		if (/^\s*$/.test(value))
			value = undefined
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [event.currentTarget.name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[event.currentTarget.name]: value,
					[event.currentTarget.name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	render() {
		return (
			<ExpansionPanel defaultExpanded CollapseProps={{timeout: 100}} >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Vulnerability, Resistance, and Immunity</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						<Grid item xs={12}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="DamageVulnerabilities">Damage Vulnerabilities</InputLabel>
								<Input
									id='DamageVulnerabilities'
									type='text'
									value={this.state.DamageVulnerabilities || ''}
									name='DamageVulnerabilities'
									onChange={this.handleResistancesChange}
									aria-describedby="DamageVulnerabilities-helper-text" />
								<FormHelperText error id="DamageVulnerabilities-helper-text">{this.state.DamageVulnerabilitiesError}</FormHelperText>
								<Tooltip disableFocusListener title={"If a creature or an object has vulnerability to a damage type, damage of that type is doubled against it."}>
										<Help className="tooltip">Hover for More About Damage Vulnerabilities</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={12}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="DamageResistances">Damage Resistances</InputLabel>
								<Input
									id='DamageResistances'
									type='text'
									value={this.state.DamageResistances || ''}
									name='DamageResistances'
									onChange={this.handleResistancesChange}
									aria-describedby="DamageResistances-helper-text" />
								<FormHelperText error id="DamageResistances-helper-text">{this.state.DamageResistancesError}</FormHelperText>
								<Tooltip disableFocusListener title={"If a creature or an object has resistance to a damage type, damage of that type is halved against it."}>
										<Help className="tooltip">Hover for More About Damage Resistance</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={12}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="DamageImmunities">Damage Immunities</InputLabel>
								<Input
									id='DamageImmunities'
									type='text'
									value={this.state.DamageImmunities || ''}
									name='DamageImmunities'
									onChange={this.handleResistancesChange}
									aria-describedby="DamageImmunities-helper-text" />
								<FormHelperText error id="DamageImmunities-helper-text">{this.state.DamageImmunitiesError}</FormHelperText>
								<Tooltip disableFocusListener title={"Creatures can be immune to damage types. A creature immune to a damage type doesn't take damage from that type of damage."}>
										<Help className="tooltip">Hover for More About Damage Immunities</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={12}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="ConditionImmunities">Condition Immunities</InputLabel>
								<Input
									id='ConditionImmunities'
									type='text'
									value={this.state.ConditionImmunities || ''}
									name='ConditionImmunities'
									onChange={this.handleResistancesChange}
									aria-describedby="ConditionImmunities-helper-text" />
								<FormHelperText error id="ConditionImmunities-helper-text">{this.state.ConditionImmunitiesError}</FormHelperText>
								<Tooltip disableFocusListener title={"Creatures can be immune to conditions. If a creature is immune to fear, poison or illusion, it isn't affected by the non-damaging effects of power with those keywords. If a creature is immune to a condition, it cannot become afflicted by that condition."}>
										<Help className="tooltip">Hover for More About Condition Immunity</Help>
								</Tooltip>
							</FormControl>
						</Grid>
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}