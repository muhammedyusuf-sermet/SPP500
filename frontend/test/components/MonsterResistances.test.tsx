import * as React from 'react';
import { mount, ReactWrapper, shallow } from 'enzyme';

import { IMonsterResistancesProps, IMonsterResistancesState, MonsterResistances } from '../../src/renderer/components/platform/pages/view_catalog/monster/MonsterResistances';
const Joi = require('joi');
import { ValidationOptions } from 'joi';

const payloadSchema = Joi.object({
	DamageVulnerabilities: Joi.string().allow('').max(200).label('DamageVulnerabilities'),
	DamageResistances: Joi.string().allow('').max(200).label('DamageResistances'),
	DamageImmunities: Joi.string().allow('').max(200).label('DamageImmunities'),
	ConditionImmunities: Joi.string().allow('').max(200).label('ConditionImmunities'),
});
const validateOptions: ValidationOptions = {
	abortEarly: false,
	convert: true,
	allowUnknown: false,
	context: {

	}
};

////// Happy Path //////
describe('Monster Resistances', () => {

	let monsterResistancesInstance: ReactWrapper<IMonsterResistancesProps, IMonsterResistancesState, MonsterResistances>;

	describe('Provided Resistance values', () => {

		beforeEach(() => {
			monsterResistancesInstance =
				mount<MonsterResistances, IMonsterResistancesProps, IMonsterResistancesState>(
					<MonsterResistances
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							DamageResistances: 'This is test damage resistance',
							DamageImmunities: 'This is test damage immunity',
							DamageVulnerabilities: 'This is test damage vulnerabilities',
							ConditionImmunities: 'This is test condition immunities'
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterResistancesInstance).toBeDefined();
		});

		it('should have the correct string for DamageResistances', () => {
			expect(monsterResistancesInstance.find('input#DamageResistances').props().value).toEqual('This is test damage resistance');
		});

		it('should have the correct string for DamageImmunities', () => {
			expect(monsterResistancesInstance.find('input#DamageImmunities').props().value).toEqual('This is test damage immunity');
		});

		it('should have the correct string for DamageVulnerabilities', () => {
			expect(monsterResistancesInstance.find('input#DamageVulnerabilities').props().value).toEqual('This is test damage vulnerabilities');
		});

		it('should have the correct string for ConditionImmunities', () => {
			expect(monsterResistancesInstance.find('input#ConditionImmunities').props().value).toEqual('This is test condition immunities');
		});
	});

	describe('componentWillReceiveProps()', () => {
		beforeEach(() => {
			monsterResistancesInstance =
				mount<MonsterResistances, IMonsterResistancesProps, IMonsterResistancesState>(
					<MonsterResistances
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							DamageResistances: 'This is test damage resistance',
							DamageImmunities: 'This is test damage immunity',
							DamageVulnerabilities: 'This is test damage vulnerabilities',
							ConditionImmunities: 'This is test condition immunities'
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterResistancesInstance).toBeDefined();
		});

		it('renders new props', () => {
			monsterResistancesInstance.setProps({
				initial: {
					DamageResistances: 'This is test new damage resistance',
					DamageImmunities: 'This is test new damage immunity',
					DamageVulnerabilities: 'This is test new damage vulnerabilities',
					ConditionImmunities: 'This is test new condition immunities'
				}
			})
			expect(monsterResistancesInstance.find('input#DamageResistances').props().value).toEqual('This is test new damage resistance');
			expect(monsterResistancesInstance.find('input#DamageImmunities').props().value).toEqual('This is test new damage immunity');
			expect(monsterResistancesInstance.find('input#DamageVulnerabilities').props().value).toEqual('This is test new damage vulnerabilities');
			expect(monsterResistancesInstance.find('input#ConditionImmunities').props().value).toEqual('This is test new condition immunities');
		});

		it('renders without crashing with same props', () => {
			monsterResistancesInstance.setProps({
				initial: {
					DamageResistances: 'This is test damage resistance',
					DamageImmunities: 'This is test damage immunity',
					DamageVulnerabilities: 'This is test damage vulnerabilities',
					ConditionImmunities: 'This is test condition immunities'
				}
			});
			expect(monsterResistancesInstance.find('input#DamageResistances').props().value).toEqual('This is test damage resistance');
			expect(monsterResistancesInstance.find('input#DamageImmunities').props().value).toEqual('This is test damage immunity');
			expect(monsterResistancesInstance.find('input#DamageVulnerabilities').props().value).toEqual('This is test damage vulnerabilities');
			expect(monsterResistancesInstance.find('input#ConditionImmunities').props().value).toEqual('This is test condition immunities');
		});
	});

	describe('Small snapshot', () => {
		it ('matches snapshot', () => {
			const shallowMonsterResistancesInstance =
				shallow<MonsterResistances, IMonsterResistancesProps, IMonsterResistancesState>(
					<MonsterResistances
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}}/>
				);
			expect(shallowMonsterResistancesInstance).toMatchSnapshot();
		});
	});

	describe('Happy Path', () => {

		beforeEach(() => {
			monsterResistancesInstance =
				mount<MonsterResistances, IMonsterResistancesProps, IMonsterResistancesState>(
					<MonsterResistances
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterResistancesInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterResistancesInstance.find('input#DamageVulnerabilities').simulate('change', { target: { value: 'Everything' } })
			monsterResistancesInstance.find('input#DamageResistances').simulate('change', { target: { value: 'None at all' } })
			monsterResistancesInstance.find('input#DamageImmunities').simulate('change', { target: { value: 'Nada' } })
			monsterResistancesInstance.find('input#ConditionImmunities').simulate('change', { target: { value: 'Nothing' } })
			expect(monsterResistancesInstance.state()).toEqual({
				// data
				DamageVulnerabilities: 'Everything',
				DamageResistances: 'None at all',
				DamageImmunities: 'Nada',
				ConditionImmunities: 'Nothing',
				// errors
				DamageVulnerabilitiesError: undefined,
				DamageResistancesError: undefined,
				DamageImmunitiesError: undefined,
				ConditionImmunitiesError: undefined
			});
		});
		// STATE CHANGE
		it('should change only DamageVulnerabilities when DamageVulnerabilities is changed', () => {
			monsterResistancesInstance.find('input#DamageVulnerabilities').simulate('change', { target: { value: 'Everything' } })
			expect(monsterResistancesInstance.state().DamageVulnerabilities).toEqual('Everything');
		})

		it('should change only DamageResistances when DamageResistances is changed', () => {
			monsterResistancesInstance.find('input#DamageResistances').simulate('change', { target: { value: 'None at all' } })
			expect(monsterResistancesInstance.state().DamageResistances).toEqual('None at all');
		})

		it('should change only DamageImmunities when DamageImmunities is changed', () => {
			monsterResistancesInstance.find('input#DamageImmunities').simulate('change', { target: { value: 'Nada' } })
			expect(monsterResistancesInstance.state().DamageImmunities).toEqual('Nada');
		})

		it('should change only ConditionImmunities when ConditionImmunities is changed', () => {
			monsterResistancesInstance.find('input#ConditionImmunities').simulate('change', { target: { value: 'Nothing' } })
			expect(monsterResistancesInstance.state().ConditionImmunities).toEqual('Nothing');
		})
		// DEFAULTS
		it('should change Type back to undefined when input is empty', () => {
			monsterResistancesInstance.find('input#DamageVulnerabilities').simulate('change', { target: { value: '' } })
			expect(monsterResistancesInstance.state().DamageVulnerabilities).toEqual(undefined);
		})

		it('should change Alignment back to undefined when input is empty', () => {
			monsterResistancesInstance.find('input#DamageResistances').simulate('change', { target: { value: '' } })
			expect(monsterResistancesInstance.state().DamageResistances).toEqual(undefined);
		})

		it('should change Size back to undefined when input is empty', () => {
			monsterResistancesInstance.find('input#DamageImmunities').simulate('change', { target: { value: '' } })
			expect(monsterResistancesInstance.state().DamageImmunities).toEqual(undefined);
		})

		it('should change Race back to undefined when input is empty', () => {
			monsterResistancesInstance.find('input#ConditionImmunities').simulate('change', { target: { value: '' } })
			expect(monsterResistancesInstance.state().ConditionImmunities).toEqual(undefined);
		})
		// VALIDATION
		const moreThan200: string = "More than 200 chars\
			01234567890123456789012345678901234567890123456789\
			01234567890123456789012345678901234567890123456789\
			01234567890123456789012345678901234567890123456789\
			01234567890123456789012345678901234567890123456789"

		it('should validate DamageVulnerabilities when DamageVulnerabilities is changed', () => {
			monsterResistancesInstance.find('input#DamageVulnerabilities').simulate('change', { target: { value: moreThan200 } })
			expect(monsterResistancesInstance.state().DamageVulnerabilities).toEqual(moreThan200);
			expect(monsterResistancesInstance.find('FormHelperText#DamageVulnerabilities-helper-text').text()).toEqual('\"DamageVulnerabilities\" length must be less than or equal to 200 characters long');
		})

		it('should validate DamageResistances when DamageResistances is changed', () => {
			monsterResistancesInstance.find('input#DamageResistances').simulate('change', { target: { value: moreThan200 } })
			expect(monsterResistancesInstance.state().DamageResistances).toEqual(moreThan200);
			expect(monsterResistancesInstance.find('FormHelperText#DamageResistances-helper-text').text()).toEqual('\"DamageResistances\" length must be less than or equal to 200 characters long');
		})

		it('should validate DamageImmunities when DamageImmunities is changed', () => {
			monsterResistancesInstance.find('input#DamageImmunities').simulate('change', { target: { value: moreThan200 } })
			expect(monsterResistancesInstance.state().DamageImmunities).toEqual(moreThan200);
			expect(monsterResistancesInstance.find('FormHelperText#DamageImmunities-helper-text').text()).toEqual('\"DamageImmunities\" length must be less than or equal to 200 characters long');
		})

		it('should validate ConditionImmunities when ConditionImmunities is changed', () => {
			monsterResistancesInstance.find('input#ConditionImmunities').simulate('change', { target: { value: moreThan200 } })
			expect(monsterResistancesInstance.state().ConditionImmunities).toEqual(moreThan200);
			expect(monsterResistancesInstance.find('FormHelperText#ConditionImmunities-helper-text').text()).toEqual('\"ConditionImmunities\" length must be less than or equal to 200 characters long');
		})
	});
});