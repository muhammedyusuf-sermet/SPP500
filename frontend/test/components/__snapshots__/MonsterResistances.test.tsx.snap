// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monster Resistances Small snapshot matches snapshot 1`] = `
<WithStyles(ExpansionPanel)
  CollapseProps={
    Object {
      "timeout": 100,
    }
  }
  defaultExpanded={true}
>
  <WithStyles(ExpansionPanelSummary)
    expandIcon={<pure(ExpandMoreIcon) />}
  >
    <WithStyles(Typography)
      className="heading"
    >
      Vulnerability, Resistance, and Immunity
    </WithStyles(Typography)>
  </WithStyles(ExpansionPanelSummary)>
  <WithStyles(ExpansionPanelDetails)>
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="DamageVulnerabilities"
          >
            Damage Vulnerabilities
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="DamageVulnerabilities-helper-text"
            id="DamageVulnerabilities"
            name="DamageVulnerabilities"
            onChange={[Function]}
            type="text"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="DamageVulnerabilities-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="DamageResistances"
          >
            Damage Resistances
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="DamageResistances-helper-text"
            id="DamageResistances"
            name="DamageResistances"
            onChange={[Function]}
            type="text"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="DamageResistances-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="DamageImmunities"
          >
            Damage Immunities
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="DamageImmunities-helper-text"
            id="DamageImmunities"
            name="DamageImmunities"
            onChange={[Function]}
            type="text"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="DamageImmunities-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="ConditionImmunities"
          >
            Condition Immunities
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="ConditionImmunities-helper-text"
            id="ConditionImmunities"
            name="ConditionImmunities"
            onChange={[Function]}
            type="text"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="ConditionImmunities-helper-text"
          />
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ExpansionPanelDetails)>
</WithStyles(ExpansionPanel)>
`;
